SIGSEGV when using an authenticated http proxy

<!-- Please note, this template is for definite bugs, not requests for
support. If you need help with Tailscale, please email
support@tailscale.com. We don't provide support via Github issues. -->

**Describe the bug**
Tailscale and netcheck fail with SIGSEGV with an authenticated http proxy

**To Reproduce**
Steps to reproduce the behavior:
1. Setup http_proxy and https_proxy in /etc/default/tailscale in "http_proxy=http://user:pass@proxy:port" form
2. Run tailscale netcheck
3. See error

```
2020/12/18 18:26:50 netcheck: probePortMapServices: me 10.0.2.15 -> gw 10.0.2.2
2020/12/18 18:26:50 netcheck: sent to 68.183.90.120:3478
2020/12/18 18:26:50 netcheck: sent to 159.89.225.99:3478
[..]
2020/12/18 18:26:51 netcheck: sent to 167.71.139.179:3478
2020/12/18 18:26:51 netcheck: sent to 103.43.75.49:3478
2020/12/18 18:26:54 netcheck: hairCheck done
2020/12/18 18:26:54 netcheck: portMap done
2020/12/18 18:26:54 netcheck: netcheck: UDP is blocked, trying HTTPS
2020/12/18 18:26:54 netcheck: derphttp: CONNECT dial to derp7.tailscale.com:443: 407 Proxy authentication required
2020/12/18 18:26:54 netcheck: netcheck: measuring HTTPS latency of tok (7): invalid response status from HTTP proxy http://user:pass@proxy:8080 on CONNECT to derp7.tailscale.com:443: 407 Proxy authentication required
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x71b47c]

goroutine 73 [running]:
tailscale.com/derp/derphttp.(*Client).dialNodeUsingProxy.func2(0xc000066a80, 0x8979c0, 0xc000062400, 0xc00002a200)
	tailscale.com@v0.0.0-00010101000000-000000000000/derp/derphttp/derphttp_client.go:587 +0xdc
created by tailscale.com/derp/derphttp.(*Client).dialNodeUsingProxy
	tailscale.com@v0.0.0-00010101000000-000000000000/derp/derphttp/derphttp_client.go:582 +0x3b9

```
4. Same for tailscaled daemon

**Expected behavior**
No sigsegv and a successful connection


**Version information:**
 - Device: virtual machine
 - OS: Arch linux
 - OS version: latest
 - Tailscale version:

```
1.2.10
  tailscale commit: e480f8ddf6f7334fda1b3d0dd1b500f2f01f961b
  other commit: 7b8c9553a1c6c4ceb525e9ce924d3ac5785f5ead
  go version: go1.15.4-tsf9db43b
```

--- a/net/tshttpproxy/tshttpproxy.go
+++ b/net/tshttpproxy/tshttpproxy.go
@@ -74,6 +74,18 @@ func GetAuthHeader(u *url.URL) (string, error) {
 	if sysAuthHeader != nil {
 		return sysAuthHeader(u)
 	}
+
+	if user := u.User.Username(); user != "" {
+		pass, ok := u.User.Password()
+		if !ok {
+			return "", nil
+		}
+
+		req := &http.Request{Header: make(http.Header)}
+		req.SetBasicAuth(user, pass)
+		return req.Header.Get("Authorization"), nil
+	}
+
 	return "", nil
 }
 

--- a/net/tshttpproxy/tshttpproxy_test.go
+++ b/net/tshttpproxy/tshttpproxy_test.go
@@ -1,0 +1,49 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build !windows
+
+package tshttpproxy
+
+import (
+	"net/url"
+	"testing"
+)
+
+func TestGetAuthHeaderNoResult(t *testing.T) {
+	const proxyURL = `http://127.0.0.1:38274`
+
+	u, err := url.Parse(proxyURL)
+	if err != nil {
+		t.Fatalf("can't parse %q: %v", proxyURL, err)
+	}
+
+	ahval, err := GetAuthHeader(u)
+	if err != nil {
+		t.Fatalf("can't get auth header value: %v", err)
+	}
+
+	if ahval != "" {
+		t.Fatalf("wanted auth header value to be empty, got: %q", ahval)
+	}
+}
+
+func TestGetAuthHeaderBasicAuth(t *testing.T) {
+	const proxyURL = `http://user:password@127.0.0.1:38274`
+	const expect = `Basic dXNlcjpwYXNzd29yZA==`
+
+	u, err := url.Parse(proxyURL)
+	if err != nil {
+		t.Fatalf("can't parse %q: %v", proxyURL, err)
+	}
+
+	ahval, err := GetAuthHeader(u)
+	if err != nil {
+		t.Fatalf("can't get auth header value: %v", err)
+	}
+
+	if ahval != expect {
+		t.Fatalf("wrong auth header value: want: %q, got: %q", expect, ahval)
+	}
+}

